ui/src/features/Editor/useCanvasCopyPaste.ts

```diff
@@ -61,38 +62,73 @@ export default ({
   );
 
   const newNodeCreation = useCallback(
-    (pastedNodes: Node[]): Node[] => {
+    (pastedNodes: Node[], mousePosition?: XYPosition): Node[] => {
       const newNodes: Node[] = [];
       const parentIdMapArray: { prevId: string; newId: string }[] = [];
+      const pastedNodeIds = new Set(pastedNodes.map((node) => node.id));
+
+      let offsetX = 0;
+      let offsetY = 0;
+
+      if (mousePosition) {
+        const reactFlowPosition = {
+          x: (mousePosition.x - x) / zoom,
+          y: (mousePosition.y - y) / zoom,
+        };
+
+        let minX = Infinity;
+        let minY = Infinity;
+
+        for (const node of pastedNodes) {
+          if (!node.parentId) {
+            if (node.position.x < minX) minX = node.position.x;
+            if (node.position.y < minY) minY = node.position.y;
+          }
+        }
+
+        offsetX = reactFlowPosition.x - minX;
+        offsetY = reactFlowPosition.y - minY;
+      } else {
+        offsetX = 25;
+        offsetY = 25;
+      }
+
       for (const n of pastedNodes) {
         // if NOT a child of a batch, offset position for user's benefit
+        const newId = generateUUID();
         const newPosition = n.parentId
           ? { x: n.position.x, y: n.position.y }
-          : { x: n.position.x + 40, y: n.position.y + 20 };
+          : { x: n.position.x + offsetX, y: n.position.y + offsetY };
 
-        const newNode: Node = {
+        const newNode = {
           ...n,
-          id: generateUUID(),
+          id: newId,
           position: newPosition,
-          selected: true, // select pasted nodes
-          data: {
-            ...n.data,
-          },
+          selected: true,
+          data: { ...n.data },
         };
 
-        if (newNode.type === "batch") {
-          parentIdMapArray.push({ prevId: n.id, newId: newNode.id });
-
-          nodes.forEach((child) => {
-            if (child.parentId === n.id) {
-              const childNewNode = {
-                ...child,
-                id: generateUUID(),
-              };
+        if (n.type === "batch") {
+          parentIdMapArray.push({ prevId: n.id, newId });
+          const batchMapping = parentIdMapArray.find(
+            (mapping) => mapping.prevId === n.id,
+          );
```

Comment:
This + the if statement is a little redundant since line 112 is pushing just adding what you then are finding here.